lib/pop.md

```diff
@@ -0,0 +1,433 @@
+# POP: Pure Object Prototypes
```

Comment:
(There is a lot of preferring not to die on this hill in my position, and giving up on reasonably quick decision making in Nixpkgs, and giving up on consistency of decision making in Nixpkgs)

My impression is that a lot of extension mechanisms in Nixpkgs are like that because they try to pretend Nix is not a computationally-general-purpose programming language. POP has to come out and pop the bubble of pretence. Which is even a plus in my opinion, but raises the risks.

On the other hand, one extension mechanism that at least recognises its computational power is modules. Which are primarily used for NixOS but now sometimes also inside Nixpkgs. I think it would be nice to include them if you already compare to other extension mechanisms. Additionally, the thing that is popular to dislike about modules is their performance overhead. It's kind of worrying that the question of runtime CPU/RAM consumption is not prominently mentioned…

Reply:
I admit I never hacked modules, so I cannot comment at the moment.

Reply:
Note that I created a PR just for POP https://github.com/NixOS/nixpkgs/pull/116275